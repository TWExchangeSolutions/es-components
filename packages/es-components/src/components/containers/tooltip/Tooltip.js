import React from 'react';
import PropTypes from 'prop-types';
import Overlay from 'react-overlays/lib/Overlay';
import styled, { ThemeProvider, withTheme } from 'styled-components';

import defaultTheme from '../../theme/defaultTheme';
import Fade from '../../util/Fade';

const TooltipBase = styled.div`
  position: absolute;
`;

const TooltipInner = styled.div`
  background-color: ${props => props.theme.colors.black};
  border-radius: 2px;
  color: ${props => props.theme.colors.white};
  font-family: sans-serif;
  font-size: 12px;
  padding: 4px 8px 5px 8px;
  text-align: center;
`;

const TooltipArrowBase = styled.div`
  border-color: transparent;
  border-style: solid;
  height: 0;
  position: absolute;
  width: 0;
`;

const TooltipTop = styled(TooltipBase)`
  margin-top: -3px;
  padding: 5px 0;
`;

const TooltipRight = styled(TooltipBase)`
  margin-right: 3px;
  padding: 0 5px;
`;

const TooltipBottom = styled(TooltipBase)`
  margin-bottom: 3px;
  padding: 5px 0;
`;

const TooltipLeft = styled(TooltipBase)`
  margin-left: -3px;
  padding: 0 5px;
`;

const TooltipArrowTop = styled(TooltipArrowBase)`
  border-top-color: ${props => props.theme.colors.black};
  border-width: 5px 5px 0;
  bottom: 0;
  left: 50%;
  margin-left: -5px;
`;

const TooltipArrowRight = styled(TooltipArrowBase)`
  border-right-color: ${props => props.theme.colors.black};
  border-width: 5px 5px 5px 0;
  left: 0;
  margin-top: -5px;
  top: 50%;
`;

const TooltipArrowBottom = styled(TooltipArrowBase)`
  border-bottom-color: ${props => props.theme.colors.black};
  border-width: 0 5px 5px;
  left: 50%;
  margin-left: -5px;
  top: 0;
`;

const TooltipArrowLeft = styled(TooltipArrowBase)`
  border-left-color: ${props => props.theme.colors.black};
  border-width: 5px 0 5px 5px;
  margin-top: -5px;
  right: 0;
  top: 50%;
`;

const FadeTransition = props => (
  <Fade duration={200} opacity={0.9} withWrapper {...props} />
);

const Popup = props => {
  const { className, children, position, style } = props;
  let TooltipStyled;
  let TooltipArrow;

  switch (position) {
    case 'right':
      TooltipStyled = TooltipRight;
      TooltipArrow = TooltipArrowRight;
      break;
    case 'bottom':
      TooltipStyled = TooltipBottom;
      TooltipArrow = TooltipArrowBottom;
      break;
    case 'left':
      TooltipStyled = TooltipLeft;
      TooltipArrow = TooltipArrowLeft;
      break;
    default:
      TooltipStyled = TooltipTop;
      TooltipArrow = TooltipArrowTop;
      break;
  }

  return (
    <TooltipStyled role="tooltip" className={className} style={style}>
      <TooltipArrow />
      <TooltipInner>{children}</TooltipInner>
    </TooltipStyled>
  );
};

Popup.propTypes = {
  children: PropTypes.any.isRequired,
  className: PropTypes.string,
  position: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),
  style: PropTypes.object
};

class Tooltip extends React.Component {
  constructor(props) {
    super(props);
    this.state = { show: false };
  }

  show = () => this.setState({ show: true });

  hide = () => this.setState({ show: false });

  render() {
    return (
      <ThemeProvider theme={this.props.theme}>
        <span>
          <span
            ref={span => {
              this.toolTipTarget = span;
            }}
            onMouseEnter={this.show}
            onMouseLeave={this.hide}
          >
            {this.props.children}
          </span>

          <Overlay
            show={this.state.show}
            placement={this.props.position}
            container={document.body}
            target={props => this.toolTipTarget}
            transition={FadeTransition}
          >
            <Popup position={this.props.position}>{this.props.content}</Popup>
          </Overlay>
        </span>
      </ThemeProvider>
    );
  }
}

Tooltip.propTypes = {
  children: PropTypes.any.isRequired,
  /** The text the tooltip displays */
  content: PropTypes.node.isRequired,
  /** Set the position of the tooltip over the content */
  position: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),
  /**
   * Theme object used by the ThemeProvider,
   * automatically passed by any parent component using a ThemeProvider
   */
  theme: PropTypes.object
};

Tooltip.defaultProps = {
  position: 'top',
  theme: defaultTheme
};

export default withTheme(Tooltip);
