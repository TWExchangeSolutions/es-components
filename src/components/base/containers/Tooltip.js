import React, { Component, PropTypes } from 'react';
import { findDOMNode } from 'react-dom';
import Overlay from 'react-overlays/lib/Overlay';
import Transition from 'react-overlays/lib/Transition';
import styled, { injectGlobal } from 'styled-components';
import { colors } from '../../theme';

injectGlobal(`
  .transition-out {
    opacity: 0;
    transition: opacity 200ms linear;
  }

  .transition-in {
    opacity: .9;
  }
`);

const TooltipBase = styled.div`
  position: absolute;
`;

const TooltipInner = styled.div`
  background-color: ${colors.black};
  border-radius: 2px;
  color: ${colors.white};
  font-family: sans-serif;
  font-size: 12px;
  padding: 3px 8px;
  text-align: center;
`;

const TooltipArrowBase = styled.div`
  border-color: transparent;
  border-style: solid;
  height: 0;
  position: absolute;
  width: 0;
`;

const TooltipTop = styled(TooltipBase)`
  margin-top: -3px;
  padding: 5px 0;
`;

const TooltipRight = styled(TooltipBase)`
  margin-right: 3px;
  padding: 0 5px;
`;

const TooltipBottom = styled(TooltipBase)`
  margin-bottom: 3px;
  padding: 5px 0;
`;

const TooltipLeft = styled(TooltipBase)`
  margin-left: -3px;
  padding: 0 5px;
`;

const TooltipArrowTop = styled(TooltipArrowBase)`
  border-top-color: ${colors.black};
  border-width: 5px 5px 0;
  bottom: 0;
  left: 50%;
  margin-left: -5px;
`;

const TooltipArrowRight = styled(TooltipArrowBase)`
  border-right-color: ${colors.black};
  border-width: 5px 5px 5px 0;
  left: 0;
  margin-top: -5px;
  top: 50%;
`;

const TooltipArrowBottom = styled(TooltipArrowBase)`
  border-bottom-color: ${colors.black};
  border-width: 0 5px 5px;
  left: 50%;
  margin-left: -5px;
  top: 0;
`;

const TooltipArrowLeft = styled(TooltipArrowBase)`
  border-left-color: ${colors.black};
  border-width: 5px 0 5px 5px;
  margin-top: -5px;
  right: 0;
  top: 50%;
`;

const Fade = (props) => (
  <Transition
    className="transition-out"
    enteredClassName="transition-in"
    enteringClassName="transition-in"
    in={props.in}
    mountOnEnter="true"
    unmountOnExit="false"
    timeout="200"
    transitionAppear="true"
  >
    {props.children}
  </Transition>
);

Fade.propTypes = {
  children: PropTypes.any.isRequired,
  in: PropTypes.boolean
};

const Popup = (props) => {
  const { className, children, position, style } = props;
  let TooltipStyled;
  let TooltipArrow;

  switch (position) {
    case 'right':
      TooltipStyled = TooltipRight;
      TooltipArrow = TooltipArrowRight;
      break;
    case 'bottom':
      TooltipStyled = TooltipBottom;
      TooltipArrow = TooltipArrowBottom;
      break;
    case 'left':
      TooltipStyled = TooltipLeft;
      TooltipArrow = TooltipArrowLeft;
      break;
    default:
      TooltipStyled = TooltipTop;
      TooltipArrow = TooltipArrowTop;
      break;
  }

  return (
    <TooltipStyled className={className} style={{ ...style }}>
      <TooltipArrow />
      <TooltipInner>
        {children}
      </TooltipInner>
    </TooltipStyled>
  );
};

Popup.propTypes = {
  children: PropTypes.any.isRequired,
  className: PropTypes.string,
  position: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),
  style: PropTypes.object
};

class Tooltip extends Component {
  constructor(props) {
    super(props);

    this.state = { show: false };
  }

  show() {
    return this.setState({ show: true });
  }

  hide() {
    return this.setState({ show: false });
  }

  render() {
    return (
      <span>
        <span ref="target" onMouseEnter={() => this.show()} onMouseLeave={() => this.hide()}>
          {this.props.children}
        </span>

        <Overlay
          show={this.state.show}
          onHide={() => this.setState({ show: false })}
          placement={this.props.position}
          container={document.body}
          target={ props => findDOMNode(this.refs.target)}
          transition={Fade}
        >
          <Popup position={this.props.position}>
            {this.props.content}
          </Popup>
        </Overlay>
      </span>
    );
  }
}

Tooltip.propTypes = {
  children: PropTypes.any.isRequired,
  /** The text the tooltip displays */
  content: PropTypes.node.isRequired,
  /** Set the position of the tooltip over the content */
  position: PropTypes.oneOf(['top', 'right', 'bottom', 'left'])
};

Tooltip.defaultProps = {
  position: 'top'
};

export default Tooltip;
