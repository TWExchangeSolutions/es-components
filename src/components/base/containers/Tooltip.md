Use a ``Tooltip`` component to display a small hover tip over content.

### Basic Tooltip:
```
<div>Hover over <Tooltip content="This is the tooltip"><a href>this link</a></Tooltip> to see the tooltip in action.</div>
```

### Tooltip with HTML:

```
<div>Hover over <Tooltip content={<span>This contains <em>some</em> <strong>HTML</strong></span>} position="left"><a href>this link</a></Tooltip> to see the tooltip in action.</div>
```
