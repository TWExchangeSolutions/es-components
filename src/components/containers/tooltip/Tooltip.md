Use a `Tooltip` component to display a small hover tip over content.

### Basic Tooltip:
```
<div>Hover over <Tooltip content="This is the tooltip"><span style={{color: '#1727EE'}}>this text</span></Tooltip> to see the tooltip in action.</div>
```

### Tooltip with HTML:
```
<div>Hover over <Tooltip content={<span>This contains <em>some</em> <strong>HTML</strong></span>} position="left"><span style={{color: '#1727EE'}}>this text</span></Tooltip> to see the tooltip in action.</div>
```
