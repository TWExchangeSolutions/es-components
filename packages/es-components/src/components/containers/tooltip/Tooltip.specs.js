/* eslint-env jest */

import React from 'react';
import renderer from 'react-test-renderer';

import Tooltip from './Tooltip';

describe('tooltip component', () => {
  it('renders tooltip children', () => {
    const tree = renderer
      .create(
        <Tooltip content="This is a tooltip">
          This is the tooltip target.
        </Tooltip>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
