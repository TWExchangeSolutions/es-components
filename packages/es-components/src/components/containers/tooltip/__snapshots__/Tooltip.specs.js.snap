// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tooltip component renders tooltip children 1`] = `
<span>
  <span
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    This is the tooltip target.
  </span>
</span>
`;
